clients/drcachesim/scheduler/scheduler.h

```diff
@@ -0,0 +1,562 @@
+        virtual stream_status_t
+        stop_speculation();
+
+        // memtrace_stream_t interface:
+
+        uint64_t
+        get_record_ordinal() const override
+        {
+            return cur_ref_count_;
+        }
+        uint64_t
+        get_instruction_ordinal() const override
+        {
+            return cur_instr_count_;
+        }
+        std::string
+        get_stream_name() const override
+        {
+            return scheduler_->get_input_name(ordinal_);
+        }
+        uint64_t
+        get_last_timestamp() const override
+        {
+            return last_timestamp_;
+        }
+        uint64_t
+        get_version() const override
+        {
+            return version_;
+        }
+        uint64_t
+        get_filetype() const override
+        {
+            return filetype_;
+        }
+        uint64_t
+        get_cache_line_size() const override
+        {
+            return cache_line_size_;
+        }
+        uint64_t
+        get_chunk_instr_count() const override
+        {
+            return chunk_instr_count_;
+        }
+        uint64_t
+        get_page_size() const override
+        {
+            return page_size_;
+        }
+
+        // Supplied for better error messages from the client.
+        std::string
+        get_source_file() const
+        {
+            return cur_src_path_;
+        }
+
+    protected:
+        scheduler_tmpl_t<RecordType, ReaderType> *scheduler_ = nullptr;
+        int ordinal_ = -1;
+        int verbosity_ = 0;
+        uint64_t cur_ref_count_ = 0;
+        uint64_t cur_instr_count_ = 0;
+        uint64_t last_timestamp_ = 0;
+        std::string cur_src_path_;
+        // Remember top-level headers for the memtrace_stream_t interface.
+        uint64_t version_ = 0;
+        uint64_t filetype_ = 0;
+        uint64_t cache_line_size_ = 0;
+        uint64_t chunk_instr_count_ = 0;
+        uint64_t page_size_ = 0;
+    };
+
+    scheduler_tmpl_t()
+    {
+    }
+    scheduler_tmpl_t(int verbosity)
+        : verbosity_(verbosity)
+    {
+    }
+    virtual ~scheduler_tmpl_t()
+    {
+    }
+
+    virtual scheduler_status_t
+    init(std::vector<input_workload_t> &workload_inputs, int output_count,
+         scheduler_options_t options);
+
+    virtual stream_t *
+    get_stream(int ordinal)
+    {
+        if (ordinal < 0 || ordinal >= static_cast<int>(outputs_.size()))
+            return nullptr;
+        return &outputs_[ordinal].stream;
+    }
+
+    std::string
+    get_error_string()
+    {
+        return error_string_;
+    }
+
+protected:
+    typedef scheduler_tmpl_t<RecordType, ReaderType> sched_type_t;
+
+    struct input_info_t {
+        std::unique_ptr<ReaderType> reader;
+        std::unique_ptr<ReaderType> reader_end;
+        memref_tid_t tid;
+        // If non-empty these records should be returned before incrementing
+        // the reader.
+        std::queue<RecordType> queue;
```

Comment:
It seems this queue is used for read-ahead or for inserting records in the input stream.  Maybe add a comment to  that effect?

Reply:
Done.